node/src/p2p/swarm_manager.rs

```diff
@@ -0,0 +1,703 @@
+        manager.start_archival_node_kad_query();
+
+        Ok(manager)
+    }
+
+    pub(crate) async fn poll(&mut self) -> Result<B::ToSwarm> {
+        loop {
+            select! {
+                _ = self.peer_tracker_info_watcher.changed() => {
+                    self.peer_health_check().await;
+                }
+                _ = self.peer_health_check_interval.tick() => {
+                    self.peer_health_check().await;
+                }
+                _ = self.gc_interval.tick() => {
+                    self.peer_tracker.gc();
+                }
+                ev = self.swarm.select_next_some() => {
+                    if let Some(ev) = self.on_swarm_event(ev).await {
+                        return Ok(ev);
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) fn context<'a>(&'a mut self) -> SwarmContext<'a, B> {
+        SwarmContext {
+            behaviour: &mut self.swarm.behaviour_mut().behaviour,
+            peer_tracker: &self.peer_tracker,
+        }
+    }
+
+    fn connect(&mut self, peer_id: &PeerId, addresses: impl Into<Option<Vec<Multiaddr>>>) {
+        if self.peer_tracker.is_connected(peer_id) {
+            return;
+        }
+
+        let addresses = addresses.into().unwrap_or_default();
+
+        let dial_opts = DialOpts::peer_id(*peer_id)
+            // Tell Swarm not to dial if peer is already connected or there
+            // is an ongoing dialing.
+            .condition(PeerCondition::DisconnectedAndNotDialing);
+
+        let dial_opts = if addresses.is_empty() {
+            dial_opts.build()
+        } else {
+            dial_opts.addresses(addresses.clone()).build()
+        };
+
+        if let Err(e) = self.swarm.dial(dial_opts) {
+            if !matches!(e, DialError::DialPeerConditionFalse(_)) {
+                warn!("Failed to dial on {addresses:?}: {e}");
+            }
+        }
+    }
+
+    /// Tries to find the node via Kademlia and connect to it
+    fn find_node_and_connect(&mut self, peer_id: &PeerId) {
+        // Check if Kademlia already knows peer_id and their addresses
+        let kad_entry_exists = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .kbucket(*peer_id)
+            .map(|bucket| {
+                bucket
+                    .iter()
+                    .any(|entry| entry.node.key.preimage() == peer_id)
+            })
+            .unwrap_or(false);
+
+        if kad_entry_exists {
+            // When `connect` is called without a specified address, then
+            // swarm asks Kademlia about it.
+            self.connect(peer_id, None);
+            return;
+        }
+
+        // When Kademlia doesn't know about peer_id, then we need to
+        // initiate a `get_closest_peers` query in order to find it and
+        // their addresses.
+        //
+        // We also avoid calling `get_closest_peers` for the same peer
+        // multiple times.
```

Comment:
```suggestion
        // When Kademlia doesn't know peer_id, we need to find it and
        // its addresses using `get_closest_peers` query.

        // First, make sure there isn't already a query in progress for this peer.
```
I'd separate the comment into two, since first part concerns the entire rest of the fn, while second concerns only the block immediately after.